### ISSUE TITLE
Incorrect QC report in MS masks

### ISSUE THREAD

**Issue by Nilser3** [ivadomed/model_seg_ms_mp2rage#78]
# Description 

I was looking at the QC in MS segmentations  (SCT 6.0) of the `basel-mp2rage`  dataset SHA commit `ffe427d4d1f62832e5f3567c8ce814eeff9b9764 `

and in subject `P001 `I found an error in the QC display:

<details><summary>sct_qc axial: </summary>

```console

--
Spinal Cord Toolbox (6.0)

sct_qc -i sub-P001_UNIT1.nii.gz -s sub-P001_UNIT1_label-SC_seg.nii.gz -d sub-P001_UNIT1_lesion-manualNeuroPoly.nii.gz -p sct_deepseg_lesion -plane axial
--


*** Generate Quality Control (QC) html report ***
Image header specifies datatype 'uint16', but array is of type 'float64'. Header metadata will be overwritten to use 'float64'.
Resample images to 0.6x0.6 mm
Image header specifies datatype 'uint8', but array is of type 'int32'. Header metadata will be overwritten to use 'int32'.
QcImage: layout with Axial slice
Compute center of mass at each slice
C:\Users\nilse\sct_6.0\python\envs\venv_sct\lib\site-packages\scipy\ndimage\_measurements.py:1535: RuntimeWarning: invalid value encountered in double_scalars
  results = [sum(input * grids[dir].astype(float), labels, index) / normalizer
.\qc\region_based\sub-P001\anat\sct_deepseg_lesion\2023_09_21_112752.490349\background_img.png
Successfully generated the QC results in .\qc\_json\qc_2023_09_21_112752.490349.json

To see the results in a browser, type:
start .\qc\index.html

```
</details>

![image](https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/4dbd7894-293f-4687-a3c6-d77f0e360eab)

<details><summary>sct_qc sagittal </summary>

```console

--
Spinal Cord Toolbox (6.0)

sct_qc -i sub-P001_UNIT1.nii.gz -s sub-P001_UNIT1_label-SC_seg.nii.gz -d sub-P001_UNIT1_lesion-manualNeuroPoly.nii.gz -p sct_deepseg_lesion -plane sagittal
--


*** Generate Quality Control (QC) html report ***
Image header specifies datatype 'uint16', but array is of type 'float64'. Header metadata will be overwritten to use 'float64'.
Resample images to 0.6x0.6 mm
Image header specifies datatype 'uint8', but array is of type 'int32'. Header metadata will be overwritten to use 'int32'.
QcImage: layout with Sagittal slice
Bounding box: x=[277, 427], y=[233, 253], z=[85, 98]
Cropping the image...
Bounding box: x=[277, 427], y=[233, 253], z=[85, 98]
Cropping the image...
Bounding box: x=[277, 427], y=[233, 253], z=[85, 98]
Cropping the image...
.\qc\region_based\sub-P001\anat\sct_deepseg_lesion\2023_09_21_112842.977264\background_img.png
Successfully generated the QC results in .\qc\_json\qc_2023_09_21_112842.977264.json

To see the results in a browser, type:
start .\qc\index.html

```
</details>

![image](https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/9adbf853-6c94-4fc9-b540-3bcc0c811d90)

# FSLeyes display and histogram :
I observe in FSLeyes the lesion is not binary (All pixels have a value of 0.9999999999997671694)

![image](https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/3ee03201-5868-4a42-9c5d-7642c53fd5e6)

## I have binarized the lesions mask and now the QC is correct,

![image](https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/e3a6c1df-f30c-4e99-8c86-8eb8b6ca101a)

![image](https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/b21d8d34-f135-4a9a-aab6-6c13d2a643ff)

# Related issues
#76 

 I will verify all the data to ensure that the masks are binary

**Comment by jcohenadad** [comment]
Could it be related to https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4178? Are you using the latest dev version of SCT? If not, can you try again with latest SCT version?

**Comment by jcohenadad** [comment]
tagging @joshuacwnewton @naga-karthik maybe they have some insights

**Comment by joshuacwnewton** [comment]
> Could it be related to https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4178? Are you using the latest dev version of SCT? If not, can you try again with latest SCT version?

I notice in the log for @Nilser3's commands, it mentions "Spinal Cord Toolbox (6.0)". Given that https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4178 was fixed after the v6.0 release, I think @jcohenadad has the right suggestion here.

**Comment by Nilser3** [comment]
After checking the QC on the entire `basel-mp2rage` dataset:

- Subjects from `sub-P001` to `sub-P031` are affected by this problem that their MS lesions pixels are not binary.

-  Subjects from `sub-P033` to `sub-P250` have correct MS binary masks.

**Comment by jcohenadad** [comment]
> Subjects from sub-P001 to sub-P031 are affected by this problem that their MS lesions pixels are not binary.

actually, if the mask is soft, then it is _not_ a problem that the mask is non-binary (it _should_ be non-binary if the mask is soft)-- what is the problem is the QC report, not the ground truth masks. So, can you pls check if the mask is supposed to be soft? if not, check if it is encoded as UINT8-- if not, encode as UINT8-- also see Joshua's message and address it please

**Comment by Nilser3** [comment]
Hi @joshuacwnewton 
I have been analyzing the pixel intensities of the mask,

## Image mask and histogram

Checking the histogram (which looks quite closed)

![image](https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/7dd3415a-117b-4352-8a6c-ac56e6ebee90)

In this python code I investigate all the intensities contained in the mask, for which I check the maximum and minimum intensity value (not counting zeros).

<img src="https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/4b5555d6-1a37-47d7-9062-594571882ff1"
 width="400" />


The maximum and minimum values of the non-zero intensities are equal, I verify that the masks are binary (0 and 0.9999999997671694) and are not soft.

## Checking the image mask type:
```
--
Spinal Cord Toolbox (6.0)

sct_image -i sub-P001_UNIT1_lesion-manualNeuroPoly.nii.gz -header
--
Image header specifies datatype 'uint16', but array is of type 'float64'. Header metadata will be overwritten to use 'float64'.
sizeof_hdr      348
data_type       FLOAT64
dim             [3, 240, 256, 176, 1, 1, 1, 1]
vox_units       mm
time_units      Unknown
datatype        64
nbyper          8
bitpix          64
pixdim          [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
```
## Changing the file type 

Change the file type to UINT8 

```
--
Spinal Cord Toolbox (6.0)

sct_image -i sub-P001_UNIT1_lesion-manualNeuroPoly.nii.gz -type uint8 -o output_uint8.nii.gz
--

Image header specifies datatype 'uint16', but array is of type 'float64'. Header metadata will be overwritten to use 'float64'.
Generate output files...
```

However, when viewing this new image type UINT8 , it is just an empty image.

Maybe the problem is in the incompatibility between the image header ('uint16') and the array type ('float64'), as the message says.

**Comment by joshuacwnewton** [comment]
> The maximum and minimum values of the non-zero intensities are equal, I verify that the masks are binary (0 and 0.9999999997671694) and are not soft.

I believe this confirms that https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4178 is the problem. That issue describes a QC bug when a segmentation has voxel values that are close to `1.0` but are not `1.0`. This caused some display issues for v6.0, but was fixed in the development version of SCT.

You should not need to change the images. You should be able to upgrade SCT to the development version, and the QC will work correctly.

(Aside: It is hard for me to say why UINT8 conversion isn't working without seeing the data itself.)

**Comment by Nilser3** [comment]
Thank you @joshuacwnewton 

Indeed,
I have installed the development version of SCT  (git-master-8a49d2949a3bf2affb56501be7cffae077a134e6) and I no longer have that problem in viewing the QC,


<details><summary> sct_qc </summary>

```console

--
Spinal Cord Toolbox (git-master-8a49d2949a3bf2affb56501be7cffae077a134e6)

sct_qc -i sub-P001_UNIT1.nii.gz -s sub-P001_UNIT1_label-SC_seg.nii.gz -d sub-P001_UNIT1_lesion-manualNeuroPoly.nii.gz -p sct_deepseg_lesion -plane sagittal
--


*** Generate Quality Control (QC) html report ***
Image header specifies datatype 'uint16', but array is of type 'float64'. Header metadata will be overwritten to use 'float64'.
Resample images to 0.6x0.6 mm
Image header specifies datatype 'uint8', but array is of type 'int64'. Header metadata will be overwritten to use 'int64'.
QcImage: layout with Sagittal slice
Bounding box: x=[277, 427], y=[233, 253], z=[85, 98]
Cropping the image...
Bounding box: x=[277, 427], y=[233, 253], z=[85, 98]
Cropping the image...
Bounding box: x=[277, 427], y=[233, 253], z=[85, 98]
Cropping the image...
./qc/home/nilser/Downloads/sct_deepseg_lesion/2023_09_29_141649.094845/background_img.png
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/css/bootstrap-table.min.css ./qc/_assets/css
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/css/bootstrap.min.css ./qc/_assets/css
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/css/bootstrap.min.css.map ./qc/_assets/css
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/css/style.css ./qc/_assets/css
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/js/jquery-3.1.0.min.js ./qc/_assets/js
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/js/bootstrap.min.js ./qc/_assets/js
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/js/yaml.min.js ./qc/_assets/js
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/js/main.js ./qc/_assets/js
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/js/bootstrap-table.min.js ./qc/_assets/js
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/js/filesaver.min.js ./qc/_assets/js
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/imgs/sct_logo.png ./qc/_assets/imgs
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/imgs/f-icon.png ./qc/_assets/imgs
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/imgs/keyright.png ./qc/_assets/imgs
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/imgs/keyup.png ./qc/_assets/imgs
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/imgs/keydown.png ./qc/_assets/imgs
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/fonts/glyphicons-halflings-regular.woff2 ./qc/_assets/fonts
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/fonts/glyphicons-halflings-regular.woff ./qc/_assets/fonts
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/fonts/glyphicons-halflings-regular.svg ./qc/_assets/fonts
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/fonts/glyphicons-halflings-regular.eot ./qc/_assets/fonts
cp /home/nilser/spinalcordtoolbox/spinalcordtoolbox/reports/assets/_assets/fonts/glyphicons-halflings-regular.ttf ./qc/_assets/fonts
Successfully generated the QC results in ./qc/_json/qc_2023_09_29_141649.094845.json

To see the results in a browser, type:
xdg-open ./qc/index.html


```
</details>

![image](https://github.com/ivadomed/model_seg_ms_mp2rage/assets/77469192/463747e5-487c-4e74-be64-df71c2dc6f8b)


However I still think it could be a problem to have binary masks with pixels where the value is not 0 / 1

**Comment by jcohenadad** [comment]
> However I still think it could be a problem to have binary masks with pixels where the value is not 0 / 1

Right, because if they are supposed to be binary and they are not, then that's wrong and should be fixed-- could you please fix that @Nilser3

**Comment by Nilser3** [comment]
Fixed and merged to the `master` in [#265](https://github.com/neuropoly/data-management/issues/265)
